### DISCUSSION TITLE
Problems with SD card slot, Wifi is connecting every second time

### DISCUSSION THREAD

**Discussion by egezgin** [schreibfaul1/ESP32-audioI2S#200] [Q&A] [answered]
Hello guys,

last weekend I got a ESP32 Audio Kit V2.2 2957. It is my first ESP32 board :). I tested your library and it is working fine, good job 👍 ! Unfortunately I have problems with other parts of the board.
I tested this example:
https://github.com/schreibfaul1/ESP32-audioI2S/blob/master/examples/ESP32-ES8388/ESP32_ES8388.ino

Here are my impressions:
- Wifi is working only every second time. I implement a workaround, so if the connection fails, the device is restarting with the function ESP.restart().
- the PA_EN pin is not correct. It is defined as 19, but on my board it is only working with 21.
- the SD_CS pin is not correct for me, because 21 is already used for PA_EN.
- the SD.begin() function does not work. Do we have a wrong SPI configuration here?

Here is the specification of the board: [Click](https://docs.ai-thinker.com/_media/esp32-a1s_v2.3_specification.pdf)
Using this document I got this configuration for SPI:

#define SD_CS         GPIO_NUM_13 //SD_DATA3
// GPIOs for SPI
#define SPI_MOSI      GPIO_NUM_15 //SD_CMD
#define SPI_MISO      GPIO_NUM_2 // SD_DATA0
#define SPI_SCK       GPIO_NUM_14 //SD_CLK

My switch configuration on the board is:
1, 4 and 5 OFF
2 and 3 ON

I tried different approaches to solve the problem, but I couldn't get it work, for example:

`SPIClass spiSD(HSPI);

pinMode(SD_CS, OUTPUT);
  digitalWrite(SD_CS, LOW);
  
  spiSD.begin(SPI_SCK, SPI_MISO, SPI_MOSI);
  spiSD.setFrequency(1000000);
 
  if (!SD.begin(SD_CS, spiSD)) {
    Serial.println("Card mount failed.");
    return;
  }`

I tried it without the pinMode and digitalWrite functions too, but SD.begin is still not working.

Do you have any ideas, what is wrong here?

PS.: My SD card is a SDHC class 6 with 16 GB from Watson. I format it with FAT32.

Thanks for your effort.

Best regards
Emrah

**Comment by biologist79** [comment]
Tipp: it's not necessary to restart the ESP. Resetting WiFi.begin() ist sufficient:
https://github.com/biologist79/ESPuino/blob/master/src/Wlan.cpp#L86
However, it's an ESP32-bug.

For SD I'd recommend to use SD_MMC instead of SPI. Needs a pin less and is round about twice as fast.

**Reply by egezgin** [reply]
Thanks for your answer.

I got it to work with WiFi.begin(), but I had to add WiFi.mode(WIFI_OFF) before the WiFi.begin() too. Now it is not necessary to do a restart. Thank you 👍 .

Any hints how I have to use SD_MMC with that board? Can I just add the code from an example?

Thanks again.

**Comment by biologist79** [comment, marked as answer]
Have a look at this one:
https://github.com/espressif/arduino-esp32/blob/master/libraries/SD_MMC/examples/SDMMC_Test/SDMMC_Test.ino

You only need GPIOs 15, 14 and 2. This is sd_mmc in 1bit-mode. But's that's already connected accordingly at your board. So you don't have to care about hardware.

But in software basically this line is important: https://github.com/espressif/arduino-esp32/blob/master/libraries/SD_MMC/examples/SDMMC_Test/SDMMC_Test.ino#L175

**Reply by egezgin** [reply]
Wow, I did not expect, that this is working directly without problems 💯. Thank you very much 👍.
I would suggest to change the example in this library, because it is really misleading.

**Comment by schreibfaul1** [comment]
Hello egezgin,
thanks for your comment. It is absolutely correct that different boards require different configurations. The example refers to the OLIMEX ADF board. And I am pleased that the Audio Lib also runs on the ESP32 Audio Kit.
There is a thread about the WiFi problem here: https://github.com/espressif/arduino-esp32/issues/2501#issuecomment-850042406
I also had the problem, with the Arduino version V1.0.6 it disappeared. Of course, if the board supports SD_MMC, this is the first choice.
Best regards
Wolle
